src/hotspot/share/gc/shenandoah/shenandoahMarkBitMap.inline.hpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2020, Red Hat, Inc. and/or its affiliates.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_VM_GC_SHENANDOAH_SHENANDOAHMARKBITMAP_INLINE_HPP
+#define SHARE_VM_GC_SHENANDOAH_SHENANDOAHMARKBITMAP_INLINE_HPP
+
+#include "gc/shenandoah/shenandoahMarkBitMap.hpp"
+#include "runtime/atomic.hpp"
+#include "utilities/count_trailing_zeros.hpp"
+
+inline size_t ShenandoahMarkBitMap::address_to_index(const HeapWord* addr) const {
+  return (pointer_delta(addr, _covered.start()) << 1) >> _shift;
+}
+
+inline HeapWord* ShenandoahMarkBitMap::index_to_address(size_t offset) const {
+  return _covered.start() + ((offset >> 1) << _shift);
+}
+
+inline bool ShenandoahMarkBitMap::mark_strong(HeapWord* heap_addr, bool& marked_first) {
+  check_mark(heap_addr);
+
+  idx_t bit = address_to_index(heap_addr);
+  verify_index(bit);
+  volatile bm_word_t* const addr = word_addr(bit);
+  const bm_word_t mask = bit_mask(bit);
+  const bm_word_t mask_final = (bm_word_t)1 << (bit_in_word(bit) + 1);
+  bm_word_t old_val = load_word_ordered(addr, memory_order_conservative);
+
+  do {
+    const bm_word_t new_val = old_val | mask;
+    if (new_val == old_val) {
+      marked_first = false;
+      return false;     // Someone else beat us to it.
+    }
+    const bm_word_t cur_val = Atomic::cmpxchg(addr, old_val, new_val, memory_order_conservative);
+    if (cur_val == old_val) {
+      marked_first = (cur_val & mask_final) == 0;
+      return true;      // Success.
+    }
+    old_val = cur_val;  // The value changed, try again.
+  } while (true);
+}
+
+inline bool ShenandoahMarkBitMap::mark_final(HeapWord* heap_addr, bool& marked_first) {
```

Comment:
It looks to me that `marked_first` is always the same as the return value? If so, can we drop that out-argument?

Reply:
Yes, that is true for mark_final(), but not for mark_strong(). I am changing it as you suggested.

Reply:
No wait, maybe that's fine then. Let me think about it.

Reply:
Same, patch sent, improvements delivered. Can resolve this discussion.